File: python-bindings/tests/shared_identity_conformance.py
"""Stable tokens observed from the public Python GameId wrapper."""

from collections.abc import Mapping
from typing import Any


def observe_shared_identity(family: str, fixture: Mapping[str, Any]) -> dict[str, Any]:
    """Read public game identity/metadata and shared-runtime diagnostics through native APIs."""
    if fixture != {"request": {}} and not (
            family == "game-identity"
            and fixture
            in (
                    {"request": {"operation": "metadata"}},
                    {"request": {"operation": "details"}},
            )
    ):
        raise ValueError("unsupported shared identity request")
    if family == "runtime-access":
        from classic_shared import get_runtime_stats, is_runtime_healthy

        available, diagnostics = [], []
        for _ in range(2):
            available.append(is_runtime_healthy())
            stats = get_runtime_stats()
            diagnostics.append(stats.is_healthy and stats.worker_threads > 0)
        return {"available": available, "diagnosticsAvailable": diagnostics}
    if family != "game-identity":
        raise ValueError("unsupported shared identity family")
    from classic_shared import GameId

    games = (GameId.Fallout4, GameId.Fallout4VR, GameId.Skyrim, GameId.Starfield)
    if fixture["request"].get("operation") == "metadata":
        return {"labels": [game.display_name() for game in games]}
    if fixture["request"].get("operation") == "details":
        return {
            "games": [
                {
                    "exeName": game.exe_name(),
                    "vr": game.is_vr(),
                    "text": str(game),
                    "repr": repr(game),
                    "equalCopy": game == getattr(GameId, game.as_str()),
                    "equalOther": game == games[(index + 1) % len(games)],
                    "hashCopy": hash(game) == hash(getattr(GameId, game.as_str())),
                }
                for index, game in enumerate(games)
            ]
        }

    return {
        "tokens": [
            game.as_str()
            for game in (
                GameId.Fallout4,
                GameId.Fallout4VR,
                GameId.Skyrim,
                GameId.Starfield,
            )
        ]
    }

File: node-bindings/classic-node/__test__/shared_identity_conformance.ts
import * as classic from "../index.js";

type JsonObject = Record<string, any>;

/** Observe stable public tokens or repeated shared-runtime access diagnostics. */
export function observeSharedIdentity(family: string, fixture: JsonObject): JsonObject {
    if (family === "game-identity") {
        const games = classic.getAllGameIds();
        if (fixture.request.operation === "metadata") return {labels: games.map(game => classic.getGameName(game))};
        return {tokens: games};
    }
    if (family === "runtime-access") {
        const available: boolean[] = [];
        const diagnosticsAvailable: boolean[] = [];
        for (let attempt = 0; attempt < 2; attempt += 1) {
            available.push(classic.isRuntimeAvailable());
            const info = classic.getRuntimeInfo();
            diagnosticsAvailable.push(info.available && info.threadCount > 0);
        }
        return {available, diagnosticsAvailable};
    }
    throw new Error("unsupported shared identity family");
}
